File: app/src/main/java/com/example/editor/EditorTab.java
package com.example.editor;

import androidx.annotation.NonNull;

import java.io.File;

public class EditorTab {

    private final String id;
    private final String filePath;
    private final String fileName;
    private String content;
    private String originalContent;
    private boolean isModified;
    private int cursorPosition;
    private int line;
    private int column;
    private final String languageId;

    public EditorTab(@NonNull String filePath, @NonNull String fileName, String content, String languageId) {
        this.id = filePath;
        this.filePath = filePath;
        this.fileName = fileName;
        this.content = content != null ? content : "";
        this.originalContent = this.content;
        this.isModified = false;
        this.cursorPosition = 0;
        this.line = 1;
        this.column = 1;
        this.languageId = languageId != null ? languageId : detectLanguage(fileName);
    }

    public static String detectLanguage(String name) {
        if (name == null) return "text";
        String lower = name.toLowerCase();
        if (lower.endsWith(".html") || lower.endsWith(".htm")) return "html";
        if (lower.endsWith(".css")) return "css";
        if (lower.endsWith(".js") || lower.endsWith(".jsx")) return "javascript";
        if (lower.endsWith(".json")) return "json";
        if (lower.endsWith(".sql")) return "sql";
        if (lower.endsWith(".py")) return "python";
        if (lower.endsWith(".java")) return "java";
        if (lower.endsWith(".kt") || lower.endsWith(".kts")) return "kotlin";
        if (lower.endsWith(".md")) return "markdown";
        if (lower.endsWith(".xml")) return "xml";
        return "text";
    }

    @NonNull
    public String getId() {
        return id;
    }

    @NonNull
    public String getFilePath() {
        return filePath;
    }

    @NonNull
    public String getFileName() {
        return fileName;
    }

    public String getContent() {
        return content;
    }

    public void updateContent(String newContent) {
        this.content = newContent != null ? newContent : "";
        this.isModified = !this.content.equals(this.originalContent);
        calculateLineColumn();
    }

    public void markSaved() {
        this.originalContent = this.content;
        this.isModified = false;
    }

    public boolean isModified() {
        return isModified;
    }

    public int getCursorPosition() {
        return cursorPosition;
    }

    public void setCursorPosition(int position) {
        this.cursorPosition = Math.max(0, Math.min(position, content.length()));
        calculateLineColumn();
    }

    public int getLine() {
        return line;
    }

    public int getColumn() {
        return column;
    }

    public String getLanguageId() {
        return languageId;
    }

    public File getFile() {
        return new File(filePath);
    }

    private void calculateLineColumn() {
        int pos = Math.min(cursorPosition, content.length());
        int currentLine = 1;
        int lastLineBreak = -1;
        for (int i = 0; i < pos; i++) {
            if (content.charAt(i) == '\n') {
                currentLine++;
                lastLineBreak = i;
            }
        }
        this.line = currentLine;
        this.column = pos - lastLineBreak;
    }
}

File: app/src/main/java/com/example/editor/EditorTab.kt
package com.example.editor

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import java.io.File

class EditorTab(
    val filePath: String,
    val fileName: String,
    initialContent: String?,
    languageId: String? = null
) {
    val id: String = filePath

    var content: String by mutableStateOf(initialContent ?: "")
        private set

    var originalContent: String = content
        private set

    var isModified: Boolean by mutableStateOf(false)
        private set

    var cursorPosition: Int by mutableIntStateOf(0)

    var line: Int by mutableIntStateOf(1)
        private set

    var column: Int by mutableIntStateOf(1)
        private set

    val languageId: String = languageId ?: detectLanguage(fileName)

    fun updateContent(newContent: String) {
        this.content = newContent
        this.isModified = (this.content != this.originalContent)
        calculateLineColumn()
    }

    fun markSaved() {
        this.originalContent = this.content
        this.isModified = false
    }

    fun updateCursor(position: Int) {
        this.cursorPosition = position.coerceIn(0, content.length)
        calculateLineColumn()
    }

    val file: File
        get() = File(filePath)

    val extension: String
        get() = file.extension

    private fun calculateLineColumn() {
        val pos = cursorPosition.coerceIn(0, content.length)
        var currentLine = 1
        var lastLineBreak = -1
        for (i in 0 until pos) {
            if (content[i] == '\n') {
                currentLine++
                lastLineBreak = i
            }
        }
        this.line = currentLine
        this.column = pos - lastLineBreak
    }

    companion object {
        fun detectLanguage(name: String?): String {
            if (name == null) return "text"
            val lower = name.lowercase()
            return when {
                lower.endsWith(".html") || lower.endsWith(".htm") -> "html"
                lower.endsWith(".css") -> "css"
                lower.endsWith(".js") || lower.endsWith(".jsx") -> "javascript"
                lower.endsWith(".ts") || lower.endsWith(".tsx") -> "typescript"
                lower.endsWith(".json") -> "json"
                lower.endsWith(".sql") -> "sql"
                lower.endsWith(".py") -> "python"
                lower.endsWith(".java") -> "java"
                lower.endsWith(".kt") || lower.endsWith(".kts") -> "kotlin"
                lower.endsWith(".md") -> "markdown"
                lower.endsWith(".xml") -> "xml"
                lower.endsWith(".sh") || lower.endsWith(".bash") -> "shell"
                lower.endsWith(".properties") || lower.endsWith(".gradle") || lower.endsWith(".toml") -> "config"
                else -> "text"
            }
        }
    }
}
